File: smolvm-python/examples/basic.py
#!/usr/bin/env python3
"""Basic example of using the smolvm Python SDK."""

import asyncio

from smolvm import Machine, MachineConfig, quick_exec


async def main():
    # Quick one-off execution
    print("=== Quick Execution ===")
    result = await quick_exec(["echo", "Hello from quick_exec!"])
    print(f"Output: {result.stdout}")
    print(f"Exit code: {result.exit_code}")
    print()

    # Using Machine with context manager
    print("=== Machine Context Manager ===")
    config = MachineConfig(name="example-machine")

    async with Machine(config) as machine:
        await machine.start()

        # Run commands directly in the VM
        result = await machine.exec(["uname", "-a"])
        print(f"VM info: {result.stdout}")

        # Run multiple commands
        result = await machine.exec(["ls", "-la", "/"])
        print(f"Root directory:\n{result.stdout}")


if __name__ == "__main__":
    asyncio.run(main())

File: smolvm-node/examples/basic.ts
/**
 * Basic usage example for smolvm SDK.
 *
 * Run with: npx ts-node examples/basic.ts
 */

import { Sandbox, withSandbox, quickExec } from "../src/index.js";

async function main() {
  console.log("=== Basic Sandbox Example ===\n");

  // Example 1: Manual lifecycle management
  console.log("1. Manual sandbox management:");
  const sandbox = await Sandbox.create({ name: "basic-example" });

  try {
    // Execute a simple command
    const result = await sandbox.exec(["echo", "Hello from smolvm!"]);
    console.log(`   Output: ${result.stdout.trim()}`);
    console.log(`   Exit code: ${result.exitCode}`);
    console.log(`   Success: ${result.success}`);

    // Execute multiple commands
    const uname = await sandbox.exec(["uname", "-a"]);
    console.log(`   System: ${uname.stdout.trim()}`);
  } finally {
    await sandbox.stop();
    await sandbox.delete();
    console.log("   Sandbox cleaned up.\n");
  }

  // Example 2: Using withSandbox helper
  console.log("2. Using withSandbox helper:");
  const result = await withSandbox({ name: "helper-example" }, async (sb) => {
    return sb.exec(["hostname"]);
  });
  console.log(`   Hostname: ${result.stdout.trim()}`);
  console.log("   Sandbox auto-cleaned up.\n");

  // Example 3: Quick execution
  console.log("3. Quick execution:");
  const quickResult = await quickExec(["date"]);
  console.log(`   Date: ${quickResult.stdout.trim()}\n`);

  // Example 4: Error handling with assertSuccess
  console.log("4. Using assertSuccess:");
  try {
    const assertResult = await withSandbox(
      { name: "assert-example" },
      async (sb) => {
        const res = await sb.exec(["true"]);
        return res.assertSuccess(); // Returns this if successful
      }
    );
    console.log(`   Command succeeded with exit code: ${assertResult.exitCode}`);
  } catch (error) {
    console.log(`   Command failed: ${error}`);
  }

  console.log("\n=== Examples Complete ===");
}

main().catch(console.error);
